### PR TITLE
Feature/issue191 wipp library refactor

### PR DESCRIPTION
**Pull request by mohamedOuladi**
## What does this PR do?

Refactored WIPP-frontend into a reusable angular library (can be published as an npm package) which exports modules and components. Also, the refactored WIPP-frontend includes an app feeding the config urls and hosting the welcome page. 

## Related issues

#191

### CODE REVIEW

File: deploy/docker/VERSION [hunk 1]
```diff
@@ -1 +1 @@
-3.0.0-devel
\ No newline at end of file
+3.0.0-devel-test
```
Comment (sunnielyu) [inline deploy/docker/VERSION]: let's adopt semantic versioning here, so start at 3.1.0

File: package.json [hunk 2]
```diff
@@ -1,9 +1,9 @@
 {
   "name": "wipp-frontend",
-  "version": "3.0.0-devel",
+  "version": "3.0.0-devel-test",
```
Comment (sunnielyu) [inline package.json]: same here, 3.1.0

File: projects/wipp-frontend-app/src/app/app-routing.module.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+const appRoutes: Routes = [
+    { path: '',   redirectTo: '/images-collections', pathMatch: 'full' },
+    { path: '**', component: PageNotFoundComponent }
+  ];
+  
+  @NgModule({
+    imports: [
+      RouterModule.forRoot(
+        appRoutes,
+        { enableTracing: true } // <-- debugging purposes only
```
Comment (sunnielyu) [inline projects/wipp-frontend-app/src/app/app-routing.module.ts]: If this is for debugging only, can we configure it via the environments file?

File: projects/wipp-frontend-lib/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# WippFrontendLib
+
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/README.md:2]: Can you update this readme to be more reflective of what the lib folder contains? Maybe a table listing all the different components? Instructions on how to use those components in an application

File: projects/wipp-frontend-lib/src/assets/img/lock-closed.svg [hunk 5]
```diff
@@ -0,0 +1 @@
...ml version="1.0" ?><svg height="32px" version="1.1" viewBox="0 0 32 32" width="32px" xmlns="http://www.w3.org/2000/svg" xmlns:sketch="http://www.bohemiancoding.com/sketch/ns" xmlns:xlink="http://www.w3.org/1999/xlink"><title/><desc/><defs/><g fill="none" fill-rule="evenodd" id="Page-1" stroke="none" stroke-width="1"><g fill="#000000" id="icon-114-lock"><path d="M16,21.9146472 L16,24.5089948 C16,24.7801695 16.2319336,25 16.5,25 C16.7761424,25 17,24.7721195 17,24.5089948 L17,21.9146472 C17.5825962,21.708729 18,21.1531095 18,20.5 C18,19.6715728 17.3284272,19 16.5,19 C15.6715728,19 15,19.6715728 15,20.5 C15,21.1531095 15.4174038,21.708729 16,21.9146472 L16,21.9146472 Z M9,14.0000125 L9,10.499235 C9,6.35670485 12.3578644,3 16.5,3 C20.6337072,3 24,6.35752188 24,10.499235 L24,14.0000125 C25.6591471,14.0047488 27,15.3503174 27,17.0094776 L27,26.9905224 C27,28.6633689 25.6529197,30 23.991212,30 L9.00878799,30 C7.34559019,30 6,28.652611 6,26.9905224 L6,17.0094776 C6,15.339581 7.34233349,14.0047152 9,14.0000125 L9,14.0000125 L9,14.0000125 Z M12,14 L12,10.5008537 C12,8.0092478 14.0147186,6 16.5,6 C18.9802243,6 21,8.01510082 21,10.5008537 L21,14 L12,14 L12,14 L12,14 Z" id="lock"/></g></g></svg>
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/assets/img/lock-closed.svg]: Are these the same icons loaded into the assets folder of the app? Do they need to be in both places? Can we remove them from the app assets?

File: projects/wipp-frontend-lib/src/lib/csv-collection/csv-collection.service.ts [hunk 6]
```diff
@@ -1,21 +1,27 @@
...
+import { WippFrontendLibConfigurationProvider } from '../wipp-frontend-lib-configuration';
 
 @Injectable({
   providedIn: 'root'
 })
 export class CsvCollectionService implements DataService<CsvCollection, PaginatedCsvCollections> {
 
-  private csvCollectionUrl = environment.apiRootUrl + '/csvCollections';
+  private csvCollectionUrl = this.apiRootUrl + '/csvCollections';
+  private plotsUiUrl = this.configurationProvider.config.plotsUiUrl;
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/csv-collection/csv-collection.service.ts:18]: Can we pass both variables through the same provider?

File: projects/wipp-frontend-lib/src/lib/generic-data/generic-data-detail/generic-data-detail.component.html [hunk 7]
```diff
@@ -11,13 +11,14 @@ <h3 class="wipp-title">Generic Data Detail</h3>
...
-        <dd class="col-md-8" *ngIf="job; else not_applicable">
+        <dd> Yoo
+        <!-- <dd class="col-md-8" *ngIf="job; else not_applicable">
           <a (click)="displayJobModal(job.id)"
              class="wipp-link"
              ngbTooltip="Show Job"
              placement="left">
             {{job.name}}
-          </a>
+          </a> -->
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/generic-data/generic-data-detail/generic-data-detail.component.html]: Is this test code? Unused commented code should be removed.

File: projects/wipp-frontend-lib/src/lib/generic-data/generic-data.service.ts [hunk 8]
```diff
@@ -59,7 +62,7 @@ export class GenericDataService implements DataService<GenericData, PaginatedGen
       }));
   }
 
-  getJob(jobUrl: string): Observable<Job> {
-    return this.http.get<Job>(jobUrl);
-  }
+  // getJob(jobUrl: string): Observable<Job> {
+  //   return this.http.get<Job>(jobUrl);
+  // }
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/generic-data/generic-data.service.ts]: Remove commented code

File: projects/wipp-frontend-lib/src/lib/images-collection/images-collection-detail/images-collection-detail.component.ts [hunk 9]
```diff
@@ -14,7 +14,7 @@ import {InlineEditorModule} from '@qontu/ngx-inline-editor';
 import {JobDetailComponent} from '../../job/job-detail/job-detail.component';
 import {Job} from '../../job/job';
 import urljoin from 'url-join';
-import {AppConfigService} from '../../app-config.service';
+//import {AppConfigService} from '../../app-config.service';
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/images-collection/images-collection-detail/images-collection-detail.component.ts]: Remove commented code

File: projects/wipp-frontend-lib/src/lib/images-collection/images-collection-detail/images-collection-detail.component.ts [hunk 10]
```diff
@@ -153,7 +152,8 @@ export class ImagesCollectionDetailComponent implements OnInit, AfterViewInit {
       map(imagesCollection => {
         this.imagesCollection = imagesCollection;
         if (this.imagesCollection.sourceCatalog) {
-          this.sourceCatalogLink = urljoin(this.appConfigService.getConfig().catalogUiUrl, this.imagesCollection.sourceCatalog);
+          //this.sourceCatalogLink = urljoin(this.appConfigService.getConfig().catalogUiUrl, this.imagesCollection.sourceCatalog);
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/images-collection/images-collection-detail/images-collection-detail.component.ts]: Remove commented code

File: projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts [hunk 11]
```diff
@@ -5,7 +5,8 @@ import {NgbModal} from '@ng-bootstrap/ng-bootstrap';
 import {NotebookService} from '../notebook.service';
 import 'prismjs';
 import * as Prism from 'prismjs';
-import * as Marked from 'marked';
+//import * as Marked from 'marked';
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts]: remove commented code

File: projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts [hunk 12]
```diff
@@ -21,6 +22,8 @@ import {HttpClient} from '@angular/common/http';
 import {NgxSpinnerService} from 'ngx-spinner';
 import {ModalErrorComponent} from '../../modal-error/modal-error.component';
 import {PlatformLocation} from '@angular/common';
+declare var nb: any;
+const marked = markedImported;
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts:30]: why not just import as marked?

File: projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts [hunk 13]
```diff
@@ -21,6 +22,8 @@ import {HttpClient} from '@angular/common/http';
 import {NgxSpinnerService} from 'ngx-spinner';
 import {ModalErrorComponent} from '../../modal-error/modal-error.component';
 import {PlatformLocation} from '@angular/common';
+declare var nb: any;
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts:27]: Can you add a comment describing why this declaration is needed?

File: projects/wipp-frontend-lib/src/lib/tensorflow-model/tensorflow-model-detail/tensorflow-model-detail.component.ts [hunk 14]
```diff
@@ -5,7 +5,7 @@ import {NgbModal} from '@ng-bootstrap/ng-bootstrap';
 import {TensorflowModelService} from '../tensorflow-model.service';
 import {TensorboardLogs, TensorflowModel} from '../tensorflow-model';
 import {JobDetailComponent} from '../../job/job-detail/job-detail.component';
-import {AppConfigService} from '../../app-config.service';
+//import {AppConfigService} from '../../app-config.service';
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/tensorflow-model/tensorflow-model-detail/tensorflow-model-detail.component.ts]: Remove commented code

File: projects/wipp-frontend-lib/src/lib/tensorflow-model/tensorflow-model-template/tensorboard-logs-template.component.ts [hunk 15]
```diff
@@ -1,7 +1,7 @@
 import {DynamicComponent} from '../../dynamic-content/dynamic.component';
 import {TensorflowModelService} from '../tensorflow-model.service';
 import {Component, OnInit} from '@angular/core';
-import {AppConfigService} from '../../app-config.service';
+//import {AppConfigService} from '../../app-config.service';
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/tensorflow-model/tensorflow-model-template/tensorboard-logs-template.component.ts]: Remove commented code

File: projects/wipp-frontend-lib/src/lib/workflow/workflow-detail/workflow-detail.component.ts [hunk 16]
```diff
@@ -11,7 +11,7 @@ import {Job} from '../../job/job';
 import {FormProperty, PropertyGroup} from 'ngx-schema-form/lib/model/formproperty';
 import {ModalErrorComponent} from '../../modal-error/modal-error.component';
 import {NgxSpinnerService} from 'ngx-spinner';
-import {AppConfigService} from '../../app-config.service';
+//import {AppConfigService} from '../../app-config.service';
```
Comment (sunnielyu) [inline projects/wipp-frontend-lib/src/lib/workflow/workflow-detail/workflow-detail.component.ts]: Remove commented code

**Review by sunnielyu — CHANGES_REQUESTED** [review]

File: projects/wipp-frontend-app/src/app/app-routing.module.ts (hunk 3 above)
Comment (mohamedOuladi) [reply projects/wipp-frontend-app/src/app/app-routing.module.ts]: this used to be part of the original code, we could move it to the environments file though

File: projects/wipp-frontend-lib/src/assets/img/lock-closed.svg (hunk 5 above)
Comment (mohamedOuladi) [reply projects/wipp-frontend-lib/src/assets/img/lock-closed.svg]: I don't think we need them in both places. We can remove the assets folder of the lib.

File: projects/wipp-frontend-lib/src/lib/csv-collection/csv-collection.service.ts (hunk 6 above)
Comment (mohamedOuladi) [reply projects/wipp-frontend-lib/src/lib/csv-collection/csv-collection.service.ts:18]: the apiRootUrl is coming from environments.ts whereas plotsUiUrl is imported trough the configuration provider  from the config,json file

File: projects/wipp-frontend-lib/src/lib/generic-data/generic-data.service.ts (hunk 8 above)
Comment (mohamedOuladi) [reply projects/wipp-frontend-lib/src/lib/generic-data/generic-data.service.ts]: this should be uncommented

File: projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts (hunk 12 above)
Comment (mohamedOuladi) [reply projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts:30]: Importing as marked caused an error "marked not found", this workaround got it to work.

File: projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts (hunk 13 above)
Comment (mohamedOuladi) [reply projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts:27]: the nb variable used to be decalred in a typing.d.ts file along with the WDZT variable. When trying to use the typing.d.ts, both variables were causing not found errors. The only way I was able to make it work was by declaring both variables inside the components. I will add a comment explaining it.

File: projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts (hunk 12 above)
Comment (sunnielyu) [reply projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts:30]: In that case, do we need to rename the variable? Can you just use `markedImport` directly?

**Review by sunnielyu — APPROVED** [review]

File: projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts (hunk 12 above)
Comment (mohamedOuladi) [reply projects/wipp-frontend-lib/src/lib/notebook/notebook-detail/notebook-detail.component.ts:30]: Renaming the variable doesn't solve the issue either. To be precise this is the error I get during the build:

`Cannot call a namespace ('markedImported')
Error: Cannot call a namespace ('markedImported')`

This issue seems to be related to the latest versions of Angular.

Here's some few links with similar issues:
https://discourse.algolia.com/t/error-cannot-call-a-namespace-algoliasearch/6632
https://stackoverflow.com/questions/59735280/angular-8-moment-error-cannot-call-a-namespace-moment

**Review by sunnielyu — APPROVED** [review]

**Pull request merged**